zookeeper-server/src/main/java/org/apache/zookeeper/server/quorum/QuorumCnxManager.java

```diff
@@ -648,17 +650,16 @@ synchronized private boolean connectOne(long sid, InetSocketAddress electionAddr
         try {
             LOG.debug("Opening channel to server " + sid);
             if (self.isSslQuorum()) {
-                 SSLSocket sslSock = self.getX509Util().createSSLSocket();
-                 setSockOpts(sslSock);
-                 sslSock.connect(electionAddr, cnxTO);
-                 sslSock.startHandshake();
```

Comment:
so the core of this fix is that you are assigning _sslSock_ to _sock_  before calling **startHandshake**
this way in case of error we are closing properly the SSLSocket

Am I correct ?

Reply:
Yes. During exception handling in the existing code, it will check and only close if the sock instance is not null.

Previously, we only assign the sock instance when it finished handshake, which may leak the connection in case if failed there after connection is created.